### PR TITLE
Add handler parameter to pushgateway functions to permit HTTP AUTH etc.

### PR DESCRIPTION
**Pull request by tim-seoss**
First stab at a handler implementation.  No tests yet, please let me know what you think.  Documenting a BASIC AUTH handler would probably be a good idea as per #60

### CODE REVIEW

**Comment by tim-seoss** [comment]
Bah, trying to rush it at the end of the day.  Will fix Sunday if I get a chance.  Sorry for the noise, please ignore for now...

File: prometheus_client/exposition.py [hunk 1]
```diff
@@ -188,10 +225,14 @@ def _use_gateway(method, gateway, job, registry, grouping_key, timeout):
...
-    if resp.code >= 400:
-        raise IOError("error talking to pushgateway: {0} {1}".format(
-            resp.code, resp.msg))
+    if handler is None:
+        resp = build_opener(handler).open(request, timeout=timeout)
+        if resp.code >= 400:
+            raise IOError("error talking to pushgateway: {0} {1}".format(
+                resp.code, resp.msg))
+    else:
+        handler(url=url, method=lambda: method, timeout=timeout,
```
Comment (brian-brazil) [inline prometheus_client/exposition.py]: Why is there a lambda for the method?

File: prometheus_client/exposition.py [hunk 2]
```diff
@@ -130,13 +130,37 @@ def push_to_gateway(gateway, job, registry, grouping_key=None, timeout=None):
                    Defaults to None
     `timeout` is how long push will attempt to connect before giving up.
               Defaults to None
+    `handler` is an optional function which can be provided to perform
+              requests to the 'gateway'.
+              Defaults to None, in which case an http or https request
+              will be carried out by a default handler.
+              If not None, the argument must be a function which accepts
+              the following arguments:
+              url, method, timeout, headers, and content
```
Comment (brian-brazil) [inline prometheus_client/exposition.py]: It'd be good to include an example one doing something simple like basic auth

File: prometheus_client/exposition.py [hunk 3]
```diff
@@ -182,10 +182,14 @@ def _use_gateway(method, gateway, job, registry, grouping_key, timeout):
     request = Request(url, data=data)
     request.add_header('Content-Type', CONTENT_TYPE_LATEST)
     request.get_method = lambda: method
-    resp = build_opener(HTTPHandler).open(request, timeout=timeout)
-    if resp.code >= 400:
-        raise IOError("error talking to pushgateway: {0} {1}".format(
-            resp.code, resp.msg))
+    if handler is None:
+        resp = build_opener(handler).open(request, timeout=timeout)
```
Comment (rossigee) [inline prometheus_client/exposition.py]: That's changed HTTPHandler to handler, which will be None. Is this correct?

Comment (brian-brazil) [inline prometheus_client/exposition.py]: That looks off alright. Can you send a PR?

Comment (rossigee) [inline prometheus_client/exposition.py]: Sent in PR #126.

**Comment by brian-brazil** [comment]
Superseded by #126

**Pull request closed without merging**